feat: add third practice project (#83)
ShaunSHamilton/web3: curriculum/locales/english/learn-digital-signatures-by-building-a-wallet.md

### --description--

For this project, you will work in the `learn-digital-signatures-by-building-a-wallet` folder. Open a bash terminal and use `cd` to change into it.

> If the tests don't pass, close any open **bash** terminals, open a new one, and run the command again.

Reviewer: This is a bit dangerous, because the freeCodeCamp server is one of those terminals

Author: I can reword these to make it a little better

```js
await new Promise(res => setTimeout(res, 1000));
const fileContents = await __helpers.getFile('learn-digital-signatures-by-building-a-wallet/generate-wallet.js');
const babelised = await __helpers.babeliser(fileContents);
const imports = babelised.getImportDeclarations();
const myImport = imports?.find(i => i.source?.value === 'elliptic');
const ecImport = myImport?.specifiers?.find(s => s.local?.name === 'EC');
assert.equal(ecImport?.local?.name, 'EC');
```

## 9

### --description--

Below that, create a new `const ec` variable, and set it to `new EC.ec('p192');`. `p192` is the type of elliptic curve you will use to generate the keys for your wallets.

### --tests--

You should have `const ec = new EC.ec('p192');` at the bottom of your `generate-wallet.js` file

```js
await new Promise(res => setTimeout(res, 1000));
const fileContents = await __helpers.getFile('learn-digital-signatures-by-building-a-wallet/generate-wallet.js');
assert.match(fileContents, /const\s*ec\s*=\s*new\s+EC\s*\.\s*ec\s*\(\s*('|"|`)p192\1\s*\)\s*;?\s*$/);
```

### --seed--

#### --"generate-wallet.js"--

```js
import EC from 'elliptic';

```

## 10

### --description--

> A keypair is generated using information from the graph of the elliptic curve. Create keypair by adding a `const keyPair` variable and setting it to `ec.genKeyPair()`.

Reviewer: ```suggestion
A keypair is generated using information from the graph of the elliptic curve. Create a keypair by adding a `const keyPair` variable and setting it to `ec.genKeyPair()`.
```

### --tests--

You should have ``console.log(`Public Key: ${publicKey}`);`` at the bottom of your `generate-wallet.js` file

```js
await new Promise(res => setTimeout(res, 1000));
const fileContents = await __helpers.getFile('learn-digital-signatures-by-building-a-wallet/generate-wallet.js');
assert.match(fileContents, /console\s*\.\s*log\s*\(\s*`Public Key: \${\s*publicKey\s*}`\s*\)\s*;?\s*$/);
```

### --seed--

#### --"generate-wallet.js"--

```js
import EC from 'elliptic';

const ec = new EC.ec('p192');
const keyPair = ec.genKeyPair();
const publicKey = keyPair.getPublic('hex');
const privateKey = keyPair.getPrivate('hex');

```

## 14

### --description--

Add another log to print the private key in the same fashion so you can see what the keys look like.

### --tests--

You should have ``console.log(`Private Key: ${privateKey}`);`` at the bottom of your `generate-wallet.js` file

```js
await new Promise(res => setTimeout(res, 1000));
const fileContents = await __helpers.getFile('learn-digital-signatures-by-building-a-wallet/generate-wallet.js');

> assert.match(fileContents, /console\s*\.\s*log\s*\(\s*`Private Key: \${\s*privateKey\s*}`\s*\)\s*;?\s*$/);

Reviewer: What is the purpose of this check? As far as I understand it, this would prevent valid code from passing: ``...${privateKey}`); const a = 1;``

```suggestion
assert.match(fileContents, /console\s*\.\s*log\s*\(\s*`Private Key: \${\s*privateKey\s*}`\s*\)/);
```

Author: The purpose is to try and make sure they stay on the right track. I usually mention to add the code at the bottom of the file, so I can check that there's nothing after it. Mainly, because I don't want them to add it above one of the existing lines. You're correct, but I think something like that is uncommon.

const privateKey = keyPair.getPrivate('hex');

console.log(`Public Key: ${publicKey}`);

```

## 15

### --description--

Run `node generate-wallet.js` to see what happens.

### --tests--

You should run `node generate-wallet.js` in the terminal

```js
await new Promise(res => setTimeout(res, 1000));
const lastCommand = await __helpers.getLastCommand();
assert.equal(lastCommand.replace(/\s+/g, ' ').trim(), 'node generate-wallet.js');
```

The terminal output should include `Public Key:` and `Private Key:`

```js
await new Promise(res => setTimeout(res, 1000));
const output = await __helpers.getTerminalOutput();
const splitOutput = output.split('generate-wallet.js');
const lastOutput = splitOutput[splitOutput.length - 1];
assert.match(lastOutput, /Public Key:/);
assert.match(lastOutput, /Private Key:/);
```

### --seed--

#### --"generate-wallet.js"--

```js
import EC from 'elliptic';

const ec = new EC.ec('p192');
const keyPair = ec.genKeyPair();
const publicKey = keyPair.getPublic('hex');
const privateKey = keyPair.getPrivate('hex');

console.log(`Public Key: ${publicKey}`);
console.log(`Private Key: ${privateKey}`);

```

## 16

### --description--

> You can see your two keys in the console. You can freely share your public key, but you need to keep your private key a secret. You can create a new keypair when you know one of the two keys. Create a `const keyPair2` variable and set the value to `ec.keyFromPrivate(privateKey, 'hex')` to generate a keypair using the private key.

Reviewer: As far as I understand, you cannot create a keypair from just a public key?

```suggestion
You can see your two keys in the console. You can freely share your public key, but you need to keep your private key a secret. You can create a new keypair when you know the private key. Create a `const keyPair2` variable and set the value to `ec.keyFromPrivate(privateKey, 'hex')` to generate a keypair using the private key.
```

Author: hmm, my understanding was that you can still generate a keypair if you only know the public key, but you just can't find the private key from that (easily anyway). But perhaps it doesn't qualify as a keypair at that point.

You should run `node generate-wallet.js` in the terminal

```js
await new Promise(res => setTimeout(res, 1000));
const lastCommand = await __helpers.getLastCommand();
assert.equal(lastCommand.replace(/\s+/g, ' ').trim(), 'node generate-wallet.js');
```

The terminal output should include `Public Key:`, `Private Key:`, and `Signature:`

```js
await new Promise(res => setTimeout(res, 1000));
const output = await __helpers.getTerminalOutput();
const splitOutput = output.split('generate-wallet.js');
const lastOutput = splitOutput[splitOutput.length - 1];
assert.match(lastOutput, /Public Key:/);
assert.match(lastOutput, /Private Key:/);
assert.match(lastOutput, /Signature:/);
```

### --seed--

#### --"generate-wallet.js"--

```js
import EC from 'elliptic';

const ec = new EC.ec('p192');
const keyPair = ec.genKeyPair();
const publicKey = keyPair.getPublic('hex');
const privateKey = keyPair.getPrivate('hex');

console.log(`Public Key: ${publicKey}`);
console.log(`Private Key: ${privateKey}`);

const keyPair2 = ec.keyFromPublic(publicKey, 'hex');
const signature = keyPair.sign('message').toDER('hex');

console.log(`Signature: ${signature}`);

```

## 29

### --description--

> Now you have used the generated keypair to sign a piece of information. Add a `const verifiedSignature` variable and set the value to `keyPair.verify('message', signature)` to verify that the signature.

Reviewer: "to verify that the signature" is what?

Author: verified

Author: yea, I'll see if I can come up with something more specific here.

You should have `const toAddress = getWalletAddressFromName(toAddressName);` in your `add-transactions.js` file

```js
await new Promise(res => setTimeout(res, 1000));
const fileContents = await __helpers.getFile('learn-digital-signatures-by-building-a-wallet/add-transaction.js');
assert.match(fileContents, /const\s+toAddress\s*=\s*getWalletAddressFromName\s*\(\s*toAddressName\s*\)\s*;?/);
```

### --seed--

#### --"add-transaction.js"--

```js
import sha256 from 'crypto-js/sha256.js';
import { 
  writeTransactions,
  getTransactions,
  getAddressBalance,
  getWalletAddressFromName
} from './blockchain-helpers.js';

const fromAddress = process.argv[2];
const toAddressName = process.argv[3];
const amount = parseInt(process.argv[4]);

const hash = sha256(fromAddress + toAddress + amount).toString();

const newTransaction = {
  hash,
  fromAddress,
  toAddress,
  amount
}

const transactions = getTransactions();
const addressBalance = getAddressBalance(fromAddress);

if (addressBalance >= amount) {
  transactions.push(newTransaction);
  writeTransactions(transactions);
} else {
  console.log('You do not have enough funds to make that transaction')
}
```

## 72

### --description--

> Open your `transactions.json` file and run `node add-transaction Me Fred 10` in the terminal to create a new transaction.

Reviewer: ```suggestion
Open your `transactions.json` file and run `node add-transaction.js Me Fred 10` in the terminal to create a new transaction.
```

assert.equal(signProp?.key?.name, 'signature');
assert.equal(signProp?.value?.name, 'signature');
```

### --seed--

#### --"add-transaction.js"--

```js
import sha256 from 'crypto-js/sha256.js';
import EC from 'elliptic';
import { 
  writeTransactions,
  getTransactions,
  getAddressBalance,
  getWalletAddressFromName
} from './blockchain-helpers.js';

const ec = new EC.ec('p192');

const fromPrivateKey = process.argv[2];
const toAddressName = process.argv[3];
const amount = parseInt(process.argv[4]);
const toAddress = getWalletAddressFromName(toAddressName);

const fromKeyPair = ec.keyFromPrivate(fromPrivateKey, 'hex');
const fromAddress = fromKeyPair.getPublic('hex');

const hash = sha256(fromAddress + toAddress + amount).toString();

const signature = fromKeyPair.sign(hash).toDER('hex');

const newTransaction = {
  hash,
  fromAddress,
  toAddress,
  amount
}

const transactions = getTransactions();
const addressBalance = getAddressBalance(fromAddress);

if (addressBalance >= amount) {
  transactions.push(newTransaction);
  writeTransactions(transactions);
} else {
  console.log('You do not have enough funds to make that transaction')
}
```

## 83

### --description--

> Make your `transactions.json` file visible and add a transaction that send zero from `Me` to `Fred`. Make sure to use the private key for `Me`.

Reviewer: ```suggestion
Make your `transactions.json` file visible and add a transaction that sends zero from `Me` to `Fred`. Make sure to use the private key for `Me`.
```

  writeTransactions,
  getWalletAddressFromName
} from './blockchain-helpers.js';

const blockchain = getBlockchain();
const previousBlock = blockchain[blockchain.length - 1];
const transactions = getTransactions();

let nonce = 0;
let hash = sha256(nonce + previousBlock.hash + JSON.stringify(transactions)).toString();
const difficulty = 2;

while (!hash.startsWith('0'.repeat(difficulty))) {
  nonce++;
  hash = sha256(nonce + previousBlock.hash + JSON.stringify(transactions)).toString();
}

const newBlock = {
  hash,
  nonce,
  previousHash: previousBlock.hash,
  transactions
}

const rewardAddress = getWalletAddressFromName('Me');
const rewardTransaction = {
  fromAddress: null,
  toAddress: rewardAddress,
  amount: 50
}

blockchain.push(newBlock);
writeBlockchain(blockchain);
writeTransactions([rewardTransaction]);
```

## 89

### --description--

Mine two new blocks to add some tokens to your blockchain.

### --tests--

You should run `node mine-block.js` in the terminal twice

```js
await new Promise(res => setTimeout(res, 1000));
const lastCommand = await __helpers.getLastCommand();
assert.equal(lastCommand.replace(/\s+/g, ' ').trim(), 'node mine-block.js');

> ```

Reviewer: It is not a big deal, because the lesson works as expected, but the test text does not match what the test is doing.

      }
    }
  }

  return true;
}

export function writeTransactions(transactions) {
  const transactionsString = JSON.stringify(transactions, null, 2);
  writeFileSync('./transactions.json', transactionsString);
}

export function getTransactions() {
  const transactionsFile = readFileSync('./transactions.json');
  const transactions = JSON.parse(transactionsFile);
  return transactions;
}

export function getAddressBalance(address) {
  const blockchain = getBlockchain();
  let balance = 0;

  // loop through blocks
  for (let i = 1; i < blockchain.length; i++) { 
    const { transactions } = blockchain[i];

    // loop through transactions
    for (let j = 0; j < transactions.length; j++) {
      const { fromAddress, toAddress, amount } = transactions[j];

      if (fromAddress === address) {
        balance -= amount;
      }

      if (toAddress === address) {
        balance += amount;
      }
    }
  }

  return balance;
}

export function getWalletAddressFromName(name) {
  const walletsFile = readFileSync('./wallets.json');
  const wallets = JSON.parse(walletsFile);
  return wallets[name];
}
```

## 108

### --description--

> Add an `if` condition that checks if `!verifiedSignature`. If it's not valid, return `false`.

Reviewer: Just for less copy-pasta 🤷‍♂️ 

```suggestion
Add an `if` condition that checks if the signature is **not** verified. If it's not valid, return `false`.
```

Author: I originally worded it close to this, but the test checks for that specific condition so I changed it cause there's too many other ways to make that same check. I will see if I can improve it a little.
